File: 2025/Quest4.cc
// Wayne Mack
// Everybody Codes
// Quest 4: PART ONE ONLY 

#include <vector>
#include <fstream>
#include <iostream>

std::vector<int> readNumbersFromFile(const std::string& filename) {
    std::vector<int> numbers;
    std::ifstream file(filename);

    if (!file.is_open()) {
        std::cerr << "Error: Could not open file: " << filename << "\n";
        return numbers; // return empty vector
    }

    int value;
    while (file >> value) {
        numbers.push_back(value);
    }

    return numbers;
}

int getGearRPM(float speed, std::vector<int> gears, int part1) {
    if (part1 == 1) {
        for (int i = gears.size() - 1; i > 0; i--) {
            float gear2 = gears[i - 1]; float gear1 = gears[i];
            float ratio  = gear1 / gear2;
            speed = speed * ratio;
            std::cout << gears[i - 1] << " to " << gears[i] << " ratio: " << ratio << " new speed: " << speed << "\n";   
        }
    }
    else {
        for (int i = 1; i < gears.size(); i++) {
            float gear1 = gears[i - 1]; float gear2 = gears[i];
            float ratio  = gear1 / gear2;
            speed = speed * ratio;
        }
    }

    int returnThis = speed;
    return returnThis;
}

int main() {
    std::vector<int> nums = readNumbersFromFile("p1sample1.txt");
    std::vector<int> nums2 = readNumbersFromFile("p1sample2.txt");
    std::vector<int> nums3 = readNumbersFromFile("puzzle.txt");

    std::cout << "Sample 1: " << getGearRPM(2025, nums,0) << "\n";
    std::cout << "Sample 2: " << getGearRPM(2025, nums2,0) << "\n";
    std::cout << "Puzzle: " << getGearRPM(2025, nums3,0) << "\n";

    /***************************************************************
     * NUMBERS TOO BIG. PART TWO AND PART THREE DONE IN JAVASCRIPT *
     ***************************************************************/

    //std::cout << "Sample 1: " << getGearRPM(10000000000000 , nums,1) << "\n";
    //std::cout << "Sample 2: " << getGearRPM(10000000000000 , nums2,1) << "\n";
    //std::cout << "Puzzle: " << getGearRPM(10000000000000 , nums3,1) << "\n";

    return 0;
}

File: 2025/Quest4.js
// Wayne Mack
// Everybody Codes
// Quest 4

const fs = require('fs');

function readNumbersFromFile(filename) {
    // Read entire file as text
    const text = fs.readFileSync(filename, 'utf8');

    // Use regex to find all numbers (integers or decimals)
    const matches = text.match(/-?\d+(\.\d+)?/g);

    // Convert matches to actual numbers
    return matches ? matches.map(Number) : [];
}

function readNumberPairs(filename) {
    const text = fs.readFileSync(filename, 'utf8');

    const lines = text.split(/\r?\n/);

    const result = [];

    for (const line of lines) {
        const trimmed = line.trim();
        if (trimmed === "") continue; // skip empty lines

        // Split by '|'
        const parts = trimmed.split('|').map(s => s.trim());

        if (parts.length === 1) {
            // Only one number → duplicate it
            const num = Number(parts[0]);
            result.push([num, num]);
        } else if (parts.length === 2) {
            // Two numbers → use them as-is
            const num1 = Number(parts[0]);
            const num2 = Number(parts[1]);
            result.push([num1, num2]);
        } else {
            // More than two parts — optional: handle or ignore
            console.warn("Unexpected format:", line);
        }
    }

    return result;
}

function getdoubleGearRPM( speed, gears) {
    for (let i = 1;  i < gears.length; i++) {
        var gear1B = gears[i - 1][1];
        var gear2A = gears[i ][0];  
        var ratio2  = gear1B / gear2A;
        speed = speed * ratio2;
    }
    return speed;
}

function getGearRPM( speed, gears, part1) {
    if (part1 == 1) {
        for (let i = gears.length - 1; i > 0; i--) {
            var gear2 = gears[i - 1]; var gear1 = gears[i];
            var ratio  = gear1 / gear2;
            speed = speed * ratio;
        }
    }
    else {
        for (let i = 1; i < gears.length; i++) {
            var gear1 = gears[i - 1]; var gear2 = gears[i];
            var ratio  = gear1 / gear2;
            speed = speed * ratio;
        }
    }
    return speed % 1 == 0?  speed : speed + 1;
}

const sample1 = readNumbersFromFile('p1sample1.txt');
const sample2 = readNumbersFromFile('p1sample2.txt');
const input = readNumbersFromFile('puzzle.txt');
const input2 = readNumbersFromFile('puzzle2.txt');
// PART ONE

console.log("Sample 1 Part 1:", getGearRPM(2025, sample1, 0));
console.log("Sample 2 Part 1:", parseInt(getGearRPM(2025, sample2, 0)));
console.log("Input Part 1:", parseInt(getGearRPM(2025, input, 0)));

// PART TWO
console.log("Sample 1 Part 2:", parseInt(getGearRPM(10000000000000 , sample1, 1)));
console.log("Sample 2 Part 2:", parseInt(getGearRPM(10000000000000 , sample2, 1)));
console.log("Input Part 2:", parseInt(getGearRPM(10000000000000 , input2, 1)));

// Part THREE
const sample3 = readNumberPairs('part3Sample.txt');
const sample3A = readNumberPairs('part3Sample2.txt');
const puzzle3 = readNumberPairs('puzzle3.txt');

console.log("Sample 3 Part 3:", parseInt(getdoubleGearRPM(100 , sample3)));
console.log("Sample 3A Part 3:", parseInt(getdoubleGearRPM(100 , sample3A)));
console.log("Puzzle 3 Part 3:", parseInt(getdoubleGearRPM(100 , puzzle3)));
